### PR TITLE
Add developer instrumentation hooks / APIs for autologging (AutologgingEventLogger)

### PR DESCRIPTION
**Pull request by dbczumar**
## What changes are proposed in this pull request?

This PR introduces an `AutologgingEventLogger` developer API that provides logging hooks for important stages in the lifecycle of an autologged ML routine. It then instruments the patched functions yielded by `safe_patch` to call `AutologgingEventLogger` hooks in the appropriate locations.

## How is this patch tested?

New unit tests for event logging & existing unit / integration tests (to ensure that event logging isn't breaking existing autologging / other MLflow workflows)

### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/utils/autologging_utils.py [hunk 1]
```diff
@@ -510,17 +517,20 @@ def __call__(self, original, *args, **kwargs):
                 raise e
 
 
+AutologgingSession = namedtuple("AutologgingSession", ["integration", "id"])
```
Comment (dbczumar) [inline mlflow/utils/autologging_utils.py:520]: Pull this out as a "public" developer API. Folks who supply their own `AutologgingEventLogger` implementations may have to reason about its content

File: mlflow/utils/autologging_utils.py [hunk 2]
```diff
@@ -533,6 +543,213 @@ def end_session(cls):
         cls._session = None
 
 
+class AutologgingEventLogger:
```
Comment (dbczumar) [inline mlflow/utils/autologging_utils.py:549]: @apurva-koti @harupy This abstraction allows developers (e.g., Databricks engineers) to plug in their own event loggers in order to instrument autologging ML execution

**Review by apurva-koti — COMMENTED** [review]
Some comments 🍣

File: mlflow/utils/autologging_utils.py [hunk 3]
```diff
@@ -326,6 +328,11 @@ def wrapper(_autolog):
         default_params = {param.name: param.default for param in param_spec.values()}
 
         def autolog(*args, **kwargs):
+            try:
+                AutologgingEventLogger.get_logger().log_autolog_called(name, args, kwargs)
+            except Exception:  # pylint: disable=broad-except
```
[the remark this answers is not in the document]
Comment (harupy) [reply mlflow/utils/autologging_utils.py:333]: +1

File: tests/autologging/test_autologging_safety_unit.py [hunk 4]
```diff
@@ -90,6 +92,95 @@ def clean_up_leaked_runs():
...
+                function_name,
+                call_args,
+                call_kwargs,
+                exception,
+            )
+        )
+
+
+@pytest.fixture
+def test_logger():
```
[the remark this answers is not in the document]
Comment (harupy) [reply tests/autologging/test_autologging_safety_unit.py:174]: Can we rename `TestLogger` to `FakeLogger`, `DummyLogger`, or `MockLogger` (whatever sounds like it's a logger for testing) and rename this fixture to `fake_logger`, `dummy_logger` or `mock_logger`?

[the remark this answers is not in the document]
Comment (apurva-koti) [reply tests/autologging/test_autologging_safety_unit.py:174]: +1

File: mlflow/utils/autologging_utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (dbczumar) [reply mlflow/utils/autologging_utils.py:333]: +1! Will file a follow-up PR for this that adds `broad-except` to `pylintrc` and drops all inline instances

File: tests/autologging/test_autologging_safety_unit.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dbczumar) [reply tests/autologging/test_autologging_safety_unit.py:174]: Sounds great! I went with `MockEventLogger`.

File: mlflow/utils/autologging_utils.py [hunk 5]
```diff
@@ -533,6 +546,213 @@ def end_session(cls):
...
+    @staticmethod
+    def set_logger(logger):
+        """
+        Configures the `AutologgingEventLogger` instance for logging. This instance
+        is exposed via `AutologgingEventLogger.get_logger()` and callers use it to invoke
+        logging hooks (e.g., AutologgingEventLogger.get_logger().log_autolog_called()).
+
+        :param logger: The instance of `AutologgingEventLogger` to use when invoking logging hooks.
+        """
+        AutologgingEventLogger._event_logger = logger
```
[the remark this answers is not in the document]
Comment (dbczumar) [reply mlflow/utils/autologging_utils.py:590]: The goal of `AutologgingEventLogger` is to provide a way for developers (e.g., Databricks engineers working on the Machine Learning Runtime) to plug their own logging / instrumentation hooks into MLflow's autologging routines. `safe_patch` is instrumented with `AutologgingEventLogger` calls at various points in the patch execution procedure.

Accordingly, we need to tell `safe_patch` *which* hooks to use throughout the patch code path by specifying an `AutologgingEventLogger` instance (i.e. some subclass of `AutologgingEventLogger` that a developer has written). The approach taken in this PR is to use a static `set_logger()` method defined on `AutologgingEventLogger` to provide the instance to `safe_patch`, which `safe_patch` consumes via the `AutologgingEventLogger.get_logger()` method.

As an alternative, we may consider separating the hooks and the static functions into two separate objects / classes (e.g, `AutologgingEventLoggerProvider` and `AutologgingEventLogger`), though I'm not sure that it helps simplify things.

@apurva-koti Let me know if this makes things clearer, and please let me know if you have suggestions for a simpler approach. cc @harupy

**Review by dbczumar — COMMENTED** [review]
@harupy @apurva-koti Addressed your comments! Let me know if you can take another look

File: tests/autologging/test_autologging_safety_unit.py [hunk 6]
```diff
@@ -90,6 +92,95 @@ def clean_up_leaked_runs():
...
+    try:
+        prev_logger = AutologgingEventLogger.get_logger()
+        logger = TestLogger()
+        AutologgingEventLogger.set_logger(logger)
+        yield logger
```
[the remark this answers is not in the document]
Comment (dbczumar) [reply tests/autologging/test_autologging_safety_unit.py]: See https://github.com/mlflow/mlflow/pull/3870#discussion_r546207648 for clarification

File: mlflow/utils/autologging_utils.py [hunk 7]
```diff
@@ -533,6 +546,213 @@ def end_session(cls):
...
+            call_args,
+            call_kwargs,
+            exception,
+        )
+
+    def log_original_function_start(
+        self, session, patch_obj, function_name, call_args, call_kwargs
+    ):
+        """
+        Called during the execution of a patched ML API associated with an autologging integration
```
Comment (harupy) [inline mlflow/utils/autologging_utils.py:678]: Can we simply say `a patched API`?

```suggestion
        Called during the execution of a patched API associated with an autologging integration
```

Maybe it's just me but `ML API` sounds like this:

Screen Shot 2020-12-21 at 12 55 36

https://nordicapis.com/choosing-the-best-machine-learning-api-for-your-project/

Comment (dbczumar) [reply mlflow/utils/autologging_utils.py:678]: Sounds good! Neat link :). I've removed "ML"!

**Comment by dbczumar** [comment]
@harupy Thanks for the additional feedback. I've addressed the naming in docstrings (`ML API` -> `API`) and also added support / test coverage for event logging from calls to `mlflow.autolog()` (by adding the `autologging_integration` decorator with name `mlflow` to `mlflow.autolog()`). Let me know how things look!

**Review by harupy — APPROVED** [review]
LGTM :)

**Comment by dbczumar** [comment]
> LGTM :)

Thank you!

**Pull request merged**